source/common/router/scoped_rds.h

```diff
@@ -0,0 +1,203 @@
+      ScopedRoutesConfigProviderManager& config_provider_manager,
+      envoy::api::v2::core::ConfigSource rds_config_source,
+      envoy::config::filter::network::http_connection_manager::v2::ScopedRoutes::ScopeKeyBuilder
+          scope_key_builder);
+
+  ~InlineScopedRoutesConfigProvider() override = default;
+
+  const std::string& name() const { return name_; }
+
+  // Envoy::Config::ConfigProvider
+  Envoy::Config::ConfigProvider::ConfigProtoVector getConfigProtos() const override {
+    Envoy::Config::ConfigProvider::ConfigProtoVector out_protos;
+    std::for_each(config_protos_.begin(), config_protos_.end(),
+                  [&out_protos](const std::unique_ptr<const Protobuf::Message>& message) {
+                    out_protos.push_back(message.get());
+                  });
+    return out_protos;
+  }
+
+  std::string getConfigVersion() const override { return ""; }
+  ConfigConstSharedPtr getConfig() const override { return config_; }
+
+private:
+  const std::string name_;
+  ConfigConstSharedPtr config_;
+  const std::vector<std::unique_ptr<const Protobuf::Message>> config_protos_;
+  const envoy::api::v2::core::ConfigSource rds_config_source_;
+};
+
+/**
+ * All SRDS stats. @see stats_macros.h
+ */
+// clang-format off
+#define ALL_SCOPED_RDS_STATS(COUNTER)                                                              \
+  COUNTER(config_reload)                                                                           \
+  COUNTER(update_empty)
+
+// clang-format on
+
+struct ScopedRdsStats {
+  ALL_SCOPED_RDS_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+// A scoped RDS subscription to be used with the dynamic scoped RDS ConfigProvider.
+class ScopedRdsConfigSubscription : public Envoy::Config::DeltaConfigSubscriptionInstance,
+                                    Envoy::Config::SubscriptionCallbacks {
+public:
+  using ScopedRouteConfigurationMap =
+      std::map<std::string, envoy::api::v2::ScopedRouteConfiguration>;
+
+  ScopedRdsConfigSubscription(
+      const envoy::config::filter::network::http_connection_manager::v2::ScopedRds& scoped_rds,
+      const uint64_t manager_identifier, const std::string& name,
+      Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
+      ScopedRoutesConfigProviderManager& config_provider_manager);
+
+  ~ScopedRdsConfigSubscription() override = default;
+
+  const std::string& name() const { return name_; }
+
+  // Envoy::Config::ConfigSubscriptionCommonBase
+  void start() override { subscription_->start({}, *this); }
+
+  // Envoy::Config::SubscriptionCallbacks
+  void onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+                      const std::string& version_info) override;
+  void onConfigUpdate(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>&,
+                      const Protobuf::RepeatedPtrField<std::string>&, const std::string&) override {
+    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
```

Comment:
I am in the process of implementing this version for all the other xDSes over in #6879, so it's just about time for these to no longer be left unimplemented. It looks like implementing the delta version, and then implementing SotW as a conversion to the delta version, should be pretty similar to source/common/upstream/cds_api_impl.cc.

Reply:
This makes sense although I would like to defer that to a follow up PR since this is one is already fairly large.